zio-logback-http4s/src/main/scala/com/softwaremill/correlator/CorrelationIdMiddleware.scala

```diff
@@ -0,0 +1,104 @@
+package com.softwaremill.correlator
+
+import java.{util => ju}
+
+import cats.data.{Kleisli, OptionT}
+import ch.qos.logback.classic.util.LogbackMDCAdapter
+import org.http4s.util.CaseInsensitiveString
+import org.http4s.{HttpRoutes, Request, Response}
+import org.slf4j.{Logger, LoggerFactory, MDC}
+import zio.random.Random
+import zio.{FiberRef, RIO, Runtime, Task, UIO, URIO, ZIO}
+import com.github.ghik.silencer.silent
+
+object CorrelationIdMiddleware {
+  import cats.implicits._
+  import zio.interop.catz._
+
+  private[correlator] final val MdcKey: String = "cid"
+  private[correlator] final val logger: Logger = LoggerFactory.getLogger(getClass.getName)
+
+  final val defaultHeaderName: String = "X-Correlation-ID"
+
+  final val defaultLogStartRequest: (String, Request[Task]) => Task[Unit] = (cid, req) =>
+    Task(logger.debug(s"Starting request with id: $cid, to: ${req.uri.path}"))
+
+  final val defaultIdGenerator: URIO[Random, String] =
+    ZIO.accessM[Random] { random =>
+      val randomUpperCaseChar: UIO[Char] = random.random.nextInt(91 - 65).map(r => (r + 65).toChar)
+      val segment: UIO[String]           = (1 to 3).toList.traverse(_ => randomUpperCaseChar).map(_.mkString)
+      (segment, segment, segment).mapN((a, b, c) => s"$a-$b-$c")
+    }
+
+  final val getCorrelationId: Task[Option[String]] = Task(Option(MDC.get(MdcKey)))
+
+  @silent("parameter value zioMDCAdapter in method addTo is never used")
+  def addTo[R <: Random](
+    headerName: String = defaultHeaderName,
+    logStartRequest: (String, Request[Task]) => Task[Unit] = defaultLogStartRequest,
+    idGenerator: RIO[R, String] = defaultIdGenerator
+  )(service: HttpRoutes[Task])(implicit runtime: Runtime[R], zioMDCAdapter: ZioMDCAdapter): HttpRoutes[Task] =
+    Kleisli { req: Request[Task] =>
+      val cidM: Task[String] =
+        req.headers.get(CaseInsensitiveString(headerName)).fold(idGenerator.asInstanceOf[Task[String]])(_.value.pure[Task])
+
+      val setupAndService: Task[Option[Response[Task]]] =
+        for {
+          cid <- cidM
+          _   <- ZIO(MDC.put(MdcKey, cid))
+          _   <- logStartRequest(cid, req)
+          r   <- service(req).value
+        } yield r
+
+      OptionT(setupAndService.ensuring(Task(MDC.remove(MdcKey)).orElse(URIO.unit)))
```

Comment:
The Kleisli and OptionT are unnecessary with ZIO.

Reply:
I don't see how since the Http4s `HttpRoutes[Task]` is an alias to `type HttpRoutes[F[_]] = Http[OptionT[F, ?], F]` which is itself an alias to `type Http[F[_], G[_]] = Kleisli[F, Request[G], Response[G]]`

So, by reduction, `type HttpRoutes[F[_]]` should be something like this:
`Kleisli[OptionT[F, ?], Request[F], Response[F]]`

Reply:
The following code doesn't compile:

```scala
  def addTo[R <: Random](
    headerName: String = defaultHeaderName,
    logStartRequest: (String, Request[Task]) => Task[Unit] = defaultLogStartRequest,
    idGenerator: RIO[R, String] = defaultIdGenerator
  )(service: HttpRoutes[Task])(implicit zioMDCAdapter: ZioMDCAdapter): HttpRoutes[Task] = {
    req: Request[Task] =>
      val cidM: Task[String] =
        req.headers.get(CaseInsensitiveString(headerName)).fold(idGenerator.asInstanceOf[Task[String]])(_.value.pure[Task])

      val setupAndService: Task[Option[Response[Task]]] =
        for {
          cid <- cidM
          _   <- ZIO(MDC.put(MdcKey, cid))
          _   <- logStartRequest(cid, req)
          r   <- service(req).value
        } yield r

      setupAndService.ensuring(Task(MDC.remove(MdcKey)).orElse(URIO.unit)).asInstanceOf[Task[Option[Response[Task]]]]
  }
```